### PR TITLE
nj 203 - added signature date to pdf

### PR DESCRIPTION
**Pull request by jachan**
## Link to pivotal/JIRA issue
- https://github.com/newjersey/affordability-pm/issues/203#issue-2738881181

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- NJ does not plan to use PIN for FY2024, but wants Signature Date to be filled in XML and PDF
- Make `primary_signature_pin` and `spouse_signature_pin` ignored so the columns can be dropped in the future

## How to test?
- Validated using `zeus_one_dep` (married) and `devito_single` (single), clicking through the flow, and checking the following:
- Signature PIN is no longer required
- XML contains Esign Date on under `ReturnHeaderState > Filer > Primary > Date Signed`
- PDF contains Signature Dates on Page 4.

## Screenshots (for visual changes)
- Single Filer
image
image

- Married Filer
image
image

### CODE REVIEW

**Comment by jachan** [comment]
Seeing `run_annotate` test failures, would appreciate advice on what to do with the DB to get this test passing!

File: app/models/state_file_nj_intake.rb [hunk 1]
```diff
@@ -113,6 +113,7 @@
 #  index_state_file_nj_intakes_on_spouse_state_id_id   (spouse_state_id_id)
 #
 class StateFileNjIntake < StateFileBaseIntake
+  self.ignored_columns = ["primary_signature_pin", "spouse_signature_pin"]
```
[the remark this answers is not in the document]
Comment (aloverso) [reply app/models/state_file_nj_intake.rb:116]: why not drop them now?

[the remark this answers is not in the document]
Comment (jachan) [reply app/models/state_file_nj_intake.rb:116]: Needs to be two separate PRs, `ignore columns` must be deployed first in order to create a `drop_column` migration. See https://github.com/codeforamerica/vita-min/commit/a4777efe616b6c8b044fe3dece0ad81605c7b50f and https://github.com/codeforamerica/vita-min/commit/94bab3dede40ea36bb9393f977016cf0dd7969b4

File: app/models/state_file_nj_intake.rb [hunk 2]
```diff
@@ -201,8 +202,5 @@ def disqualifying_eligibility_rules
 
   def validate_state_specific_w2_requirements(w2); end
 
-  def ask_for_signature_pin?
-    true
-  end
```
Comment (aloverso) [inline app/models/state_file_nj_intake.rb:206]: [dust/pebble] does this default to false? Maybe we could keep the method but have it explicitly return false, since that is explicit logic that we want to define for NJ rather than relying on an external default

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb [hunk 3]
```diff
@@ -279,6 +279,40 @@
...
+        let(:submission) {
+          create :efile_submission, tax_return: nil, data_source: create(
+            :state_file_nj_intake,
+            primary_ssn: "123456789",
+            primary_esigned: "yes",
+            primary_esigned_at: DateTime.new(2024, 12, 19, 12)
+          )
+        }
+        it 'enters the date into the PDF signature field' do
+          expect(pdf_fields["Date1_es_:signer:date"]).to eq "2024-12-19"
```
Comment (aloverso) [inline spec/lib/pdf_filler/nj1040_pdf_spec.rb:293]: [pebble] can we expect the spouse one to be empty here?

**Review by aloverso — COMMENTED** [review]
Looks good! Left a few comments

File: app/lib/pdf_filler/nj1040_pdf.rb [hunk 4]
```diff
@@ -94,6 +94,10 @@ def hash_for_pdf
         Group246: if get_mfj_spouse_ssn
                     @xml_document.at("Body SpouCuPartPrimGubernElectFund").present? ? 'Choice1' : 'Choice2'
                   end,
+        
+        # signature fields
+        'Date1_es_:signer:date': @xml_document.at("ReturnHeaderState Filer Primary DateSigned")&.text,
+        'Date2_es_:signer:date': @xml_document.at("ReturnHeaderState Filer Secondary DateSigned")&.text
```
[the remark this answers is not in the document]
Comment (jachan) [reply app/lib/pdf_filler/nj1040_pdf.rb:100]: If the secondary filer doesn't exist, the `&` operator means the PDF field gets set to nil and the field is empty, which is what we want.

File: app/models/state_file_nj_intake.rb [hunk 5]
```diff
@@ -5,6 +5,7 @@
 #  id                                                     :bigint           not null, primary key
 #  account_number                                         :string
 #  account_type                                           :integer          default("unfilled"), not null
+#  bank_name                                              :string
```
[the remark this answers is not in the document]
Comment (jachan) [reply app/models/state_file_nj_intake.rb:8]: Hmm. Thanks, will try that.

File: app/lib/submission_builder/return_header.rb [hunk 6]
```diff
@@ -30,10 +30,10 @@ def document
             xml.TaxpayerSSN @submission.data_source.primary.ssn if @submission.data_source.primary.ssn.present?
             xml.DateOfBirth date_type(@submission.data_source.primary.birth_date) if @submission.data_source.primary.birth_date.present?
             xml.TaxpayerPIN @submission.data_source.primary_signature_pin if @submission.data_source.ask_for_signature_pin?
-            xml.DateSigned date_type_for_timezone(@submission.data_source.primary_esigned_at)&.strftime("%F") if @submission.data_source.ask_for_signature_pin?
+            xml.DateSigned date_type_for_timezone(@submission.data_source.primary_esigned_at)&.strftime("%F") if @submission.data_source.primary_esigned_yes?
```
[the remark this answers is not in the document]
Comment (mmazanec22) [reply app/lib/submission_builder/return_header.rb:33]: We could add a method like "add_esigned_at" to state intake models instead of forking return_header

File: spec/factories/state_file_nj_intakes.rb [hunk 7]
```diff
@@ -122,10 +122,6 @@
     routing_number { "011234567" }
     account_number { "123456789" }
     account_type { 1 }
-    primary_signature_pin { '12345' }
```
[the remark this answers is not in the document]
Comment (mmazanec22) [reply spec/factories/state_file_nj_intakes.rb:125]: I pulled down this PR, added this here
```
    primary_esigned_at { DateTime.new(2024, 12, 19, 12)}
    spouse_esigned_at { DateTime.new(2024, 12, 18, 12) }
    primary_esigned { 'yes' }
    spouse_esigned { 'yes' }
```
and deleted the attribute assignments in pdf spec. The tests pass. I think maybe these attributes were doing nothing because you weren't assigning primary_esigned and spouse_esigned to yes in this file.

[the remark this answers is not in the document]
Comment (jachan) [reply spec/factories/state_file_nj_intakes.rb:125]: Hm, I did try that. 

Agree that it's preferable for these to live in the factory, since all submitted intakes will need to be signed to be submitted. Will give this another shot, thanks for the comment!

File: app/lib/submission_builder/return_header.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jachan) [reply app/lib/submission_builder/return_header.rb:33]: I think `esigned_at` is already defined for all states! 

https://github.com/codeforamerica/vita-min/blob/5c77360c9d03bc0356b707bedc1adb8c649fd40a/app/forms/state_file/esign_declaration_form.rb#L33

 The shared Esign Declaration Form sets the attribute, so if any taxpayer checks the esign box, `primary_esigned_at` gets set.

File: app/models/state_file_nj_intake.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (jachan) [reply app/models/state_file_nj_intake.rb:8]: That did it! Thank you!

File: app/lib/submission_builder/return_header.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (arinchoi03) [reply app/lib/submission_builder/return_header.rb:33]: i'll ask about this!

[the remark this answers is not in the document]
Comment (arinchoi03) [reply app/lib/submission_builder/return_header.rb:33]: @jachan from our PM 

> we want to include it for all states, triggered by the filer checking that e-sign box and submitting their return! i think this is just something that was missed last year

Given this response, i think this change is great for all states 🙏

File: spec/factories/state_file_nj_intakes.rb [hunk 8]
```diff
@@ -15,6 +16,7 @@
 #  current_sign_in_ip                                     :inet
 #  current_step                                           :string
 #  date_electronic_withdrawal                             :date
+#  df_data_import_failed_at                               :datetime
```
Comment (arinchoi03) [inline spec/factories/state_file_nj_intakes.rb:19]: I think you need to extend the ignored columns from the base intakes to get this removed!

File: app/models/state_file_nj_intake.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (arinchoi03) [reply app/models/state_file_nj_intake.rb:116]: pls change to:
```
self.ignored_columns += ["primary_signature_pin", "spouse_signature_pin"]
```

**Review by arinchoi03 — COMMENTED** [review]
some comments, will re-review

File: spec/lib/pdf_filler/nj1040_pdf_spec.rb [hunk 9]
```diff
@@ -279,6 +279,41 @@
...
     end
 
+    describe "esign date" do
+      context 'filer is single' do
+        let(:submission) {
+          create :efile_submission, tax_return: nil, data_source: create(
+            :state_file_nj_intake,
+            primary_ssn: "123456789",
+            primary_esigned: "yes",
+            primary_esigned_at: DateTime.new(2024, 12, 19, 12)
```
[the remark this answers is not in the document]
Comment (jachan) [reply spec/lib/pdf_filler/nj1040_pdf_spec.rb:289]: A good question, thanks for bringing this up! In the context of this PR, we are adding tests for PDF and XML generation, so I'm thinking through them individually below. The DateTime Ruby docs show it's possible to specify a timezone for DateTime objects: https://ruby-doc.org/stdlib-2.6.1/libdoc/date/rdoc/DateTime.html#method-c-new. It defaults to UTC, which means we're currently creating DateTime with an offset of 0. 

XML Generation Tests (ALL STATES)
The DateTime object is converted to a string by this helper method:

https://github.com/codeforamerica/vita-min/blob/main/app/lib/submission_builder/formatting_methods.rb#L29

Which converts a DateTime into a date using state_code.

The existing tests use ET explicitly when checking for midnight edge cases (see lines 288 and 307 in `return_header_spec.rb`), but it's helpful to note that the max offset for the contingous 48 states in the US is UTC-8 (Hawaii is UTC-10 and Alaska is UTC-9). In all those cases, noon UTC will still be the same date when converted to local state_code date. There should be no date discrepancy here.

PDF Generation Tests (NJ ONLY)
The PDF generation takes the XML string and compares it to a static expected string, 2024-12-19. The converted datetime will always be on this day for NJ since we're converting the noon 12-19-2024 UTC timestamp to ET (UTC-5). This means that it corresponds to 12-19-2024, 7am.

Hope this answers your question, let me know if you have additional concerns.

[the remark this answers is not in the document]
Comment (jachan) [reply spec/lib/pdf_filler/nj1040_pdf_spec.rb:289]: I'm going to rewrite the return_header_spec tests to test all states DateSigned functionality in the next commit

[the remark this answers is not in the document]
Comment (jachan) [reply spec/lib/pdf_filler/nj1040_pdf_spec.rb:289]: This is proving tougher to test than expected, given that every state has a different way of indicating that a spouse exists 😬 . 

Ideally, I'd create an intake for a single filer, verify that the XML has the correct `Primary DateSigned` attribute present, and no `Secondary DateSigned` element. Then I'd create an intake for a filer with a spouse, verify the XML has correct `Primary DateSigned` and `Secondary DateSigned` attributes.

Suggestions for how to set the spouse case across all active states? It seems right now like I'd have to modify every state's intake factory to set `spouse_esigned` and `spouse_esigned_at`. I could do that in the default case, which would make things easy, but it seems very wrong to set `spouse_esigned` and `spouse_esigned_at` for the default single filer.

File: spec/factories/state_file_id_intakes.rb [hunk 10]
```diff
@@ -108,7 +108,7 @@
     end
```
Comment (arinchoi03) [inline spec/factories/state_file_id_intakes.rb:108]: https://stackoverflow.com/a/41626919
I think we're using `transient` wrong here
>transient attributes allow you to pass in data that isn’t an attribute on the model.

All these traits (except for `filing_status` which makes sense here) should not be in this transient block.

File: spec/factories/state_file_md_intakes.rb [hunk 11]
```diff
@@ -129,9 +129,10 @@
     primary_first_name { "Mary" }
     primary_middle_initial { "A" }
     primary_last_name { "Lando" }
-    primary_birth_date { Date.new(1950, 01, 01) } # matches the bday in md_minimal.json
```
Comment (arinchoi03) [inline spec/factories/state_file_md_intakes.rb:132]: this is a weird change?

Comment (jachan) [reply spec/factories/state_file_md_intakes.rb:132]: Auto-rubocop correct due to rule https://docs.rubocop.org/rubocop/cops_style.html#stylenumericliteralprefix. I have fix rubocop violations on save enabled. Strange that it assumes 01 is octal.

Changing to `1950, 1, 1` takes away the violation.

**Comment by jachan** [comment]
Closing and recreating PR to retrigger Heroku build
